### ISSUE TITLE
关于voco2d重建音频质量一般问题

### ISSUE THREAD

**Issue by RoversCode** [zai-org/GLM-TTS#12]
本人对voco2d比较感兴趣。因此简单写了一下推理脚本，在一个数据集上做了测试。

测试结果如下
pesq_wb:
  平均值: 2.9685
  最低值: 2.1793
  最高值: 3.7371
------------------------------
pesq_nb:
  平均值: 3.4786
  最低值: 2.8945
  最高值: 3.9875
------------------------------
stoi:
  平均值: 0.9462
  最低值: 0.8901
  最高值: 0.9823
------------------------------
sim_score:
  平均值: 0.9021
  最低值: 0.7715
  最高值: 0.9586
------------------------------

放一个音频例子，
【原音频】
[原音频1.wav](https://github.com/user-attachments/files/24099611/1.wav)
【重建音频】
[重建音频.wav](https://github.com/user-attachments/files/24099623/default.wav)

听一下重建音频，可以很明显听到，声音稳定性出现了问题，并伴随了一些噪声出现。

然后以下是我用Cosy2的HifiGAN，重建的指标与效果

<img width="1336" height="242" alt="Image" src="https://github.com/user-attachments/assets/fd1566c2-6deb-4148-ba6b-57cdf7fd9883" />

【同一个音频的效果】

[hifigan.wav](https://github.com/user-attachments/files/24099857/hifigan.wav)

以下是voco2d的推理代码

`import argparse
import glob
import os

import torch
import torchaudio
from tqdm import tqdm

from utils.vocos_util import load_vocos_jit, FS32K

def load_audio(path: str, target_sr: int, device: str) -> torch.Tensor:
    """
    加载音频并重采样到 target_sr，返回形状为 (1, T) 的 mono waveform（float32，-1~1）。
    """
    wav, sr = torchaudio.load(path)  # (C, T)

    # 转为单声道
    if wav.size(0) > 1:
        wav = wav.mean(dim=0, keepdim=True)

    # 重采样到 vocos 期望的采样率
    if sr != target_sr:
        resampler = torchaudio.transforms.Resample(sr, target_sr)
        wav = resampler(wav)

    # 移到目标设备
    wav = wav.to(device)

    return wav

def process_one(vocos, audio_path: str, device: str = "cuda"):
    """
    处理单个音频：返回 (recons_wav_32k, origin_wav_32k)，均为 (1, T) 的 Tensor（在 CPU 上）。
    如有错误返回 (None, None)。
    """
    try:
        wav_32k = load_audio(audio_path, target_sr=FS32K, device=device)  # (1, T) on device

        with torch.no_grad():
            mel = vocos.stft_mel(wav_32k)  # (B, n_mel, frames)
            recons = vocos(mel)           # (B, T)

        recons = recons.squeeze(0).cpu().unsqueeze(0)      # (1, T)
        origin = wav_32k.squeeze(0).cpu().unsqueeze(0)     # (1, T)

        return recons, origin
    except Exception as e:
        print(f"处理 {audio_path} 时出错：{e}")
        return None, None

def batch_reconstruct(input_dir: str, output_dir: str, device: str = "cuda"):
    """
    批量处理目录中的所有音频文件。
    在 output_dir 下生成：
      - origin/: 预处理后的原音频（32k 单声道）
      - recons/: vocos2d 重建后的音频
    """
    # 设备设置
    device = device if torch.cuda.is_available() and device.startswith("cuda") else "cpu"
    print(f"使用设备: {device}")

    # 创建输出目录
    origin_dir = os.path.join(output_dir, "origin")
    recons_dir = os.path.join(output_dir, "recons")
    os.makedirs(origin_dir, exist_ok=True)
    os.makedirs(recons_dir, exist_ok=True)

    # 加载 vocos 模型
    vocos = load_vocos_jit(device=device)

    # 收集音频文件
    exts = ["*.wav", "*.flac", "*.mp3", "*.ogg", "*.opus", "*.m4a"]
    audio_files = []
    for ext in exts:
        audio_files.extend(glob.glob(os.path.join(input_dir, ext)))

    print(f"在 {input_dir} 中找到 {len(audio_files)} 个音频文件")

    # 逐个处理
    for audio_path in tqdm(audio_files):
        filename = os.path.basename(audio_path)
        name, _ = os.path.splitext(filename)

        recons_wav, origin_wav = process_one(vocos, audio_path, device=device)
        if recons_wav is None:
            continue
        # 计算音量，并同步
        from utils.audio import calculate_decibel, adjust_audio_level
        recons_db = calculate_decibel(recons_wav)
        original_db = calculate_decibel(origin_wav)
        # 去最大的值，并同步
        max_db = max(recons_db, original_db)
        recons_wav = adjust_audio_level(recons_wav, max_db)
        origin_wav = adjust_audio_level(origin_wav, max_db)
        # 保存
        torchaudio.save(os.path.join(origin_dir, f"{name}.wav"), origin_wav, FS32K)
        torchaudio.save(os.path.join(recons_dir, f"{name}.wav"), recons_wav, FS32K)

    print(f"所有音频已处理完成，结果保存在 {output_dir}")

def main():
    parser = argparse.ArgumentParser(description="批量 Vocos2D 重建脚本")
    parser.add_argument("--input_dir", type=str, default="notebook/audios/speech/ref_audio", help="输入音频文件目录")
    parser.add_argument("--output_dir", type=str, default="test/vocos2d_recons", help="输出目录（将创建 origin/ 和 recons/）")
    parser.add_argument(
        "--device",
        type=str,
        default="cuda:7",
        help='设备，如 "cuda" 或 "cpu"，默认为 "cuda"（如果可用）',
    )
    args = parser.parse_args()

    batch_reconstruct(
        input_dir=args.input_dir,
        output_dir=args.output_dir,
        device=args.device,
    )

if __name__ == "__main__":
    main()

`

是我推理的处理没写好，还是说，voco2d效果就是这样的

**Comment by aqtq314** [comment]
vocos2d针对精品音色、以及Flow Matching生成的mel进行过finetune，泛化性会下降一些

**Comment by RoversCode** [comment]
> 仔细看了代码以及听了音频，代码没检查出问题。GT音频自带一点混响确实是vocos2d训练的时候没怎么接触过的，vocos2d的训练数据是一小半语音+一大半歌声，都是干声，后续微调也是。
> 
> 另一方面我们这边的观察是，像vocos2d以及其他中间时频谱重建步骤比重比较大的声码器普遍高音性能更稳，而且不电，不怎么会有过剩的金属感，缺点就是低音尤其气泡音容易糊成喘气声；直接重建波形的声码器低音会稳一些，其他的特性也是正好反过来。感觉这些特性在音频里也多少有点体现。

好的，感谢回复.

**Comment by JohnHerry** [comment]
vocos如果低频带上采样效果差，那么在训练时，多分辨率判别器配置时，增加几个粗分辨率频带，有助于改进效果吗？

**Comment by aqtq314** [comment]
> vocos如果低频带上采样效果差，那么在训练时，多分辨率判别器配置时，增加几个粗分辨率频带，有助于改进效果吗？

不是低频带，是指基频音高比较低的（比如男声，尤其蒙古呼麦
